### DISCUSSION TITLE
Creating a plugin that calls custom hooks in the user's gatsby-node.js file

### DISCUSSION THREAD

**Discussion by chrispytoes** [gatsbyjs/gatsby#29496] [Help]
Is there a way to retrieve the location of the main `gatsby-node.js` file from within my plugin (specifically a theme) so I can call my own custom hooks that the plugin user may define?

**Comment by KyleAMathews** [comment]
You can do something like:

```js
const path = require(`path`)

const pathToPlugin = require.resolve(`gatsby-plugin-x`)
const pathToGatsbyNodefile = path.join(pathToPlugin, `gatsby-node.js`)
```

Note, this might not work the way you expect as Gatsby restricts `gatsby-node.js` files from exporting functions other than ones matching Gatsby APIs.

**Reply by chrispytoes** [reply]
It looks like that would import my own plugin's `gatsby-node.js` file no? I'm trying to retrieve the main root project's `gatsby-node.js` file, not my plugin's file.

Is it more recommended to just accept functions in my plugin options, if that's even possible? Or should I have my plugin look for a new JS file that the user must create?

**Reply by KyleAMathews** [reply]
oh sorry, you asked about the `main` `gatsby-node.js`.

>Is it more recommended to just accept functions in my plugin options, if that's even possible? Or should I have my plugin look for a new JS file that the user must create?

Yeah the latter. We don't allow passing in functions directly as plugin options need to be serializable. But they can pass in a path to a js file (or you can just instruct them exactly what to create e.g. `./src/gatsby-plugin-x-helpers.js` and check for its existence).

**Reply by chrispytoes** [reply]
Thanks, that just leaves me with one more thing.

Is there a way to reliably find the location of the main root project from my theme? Since my theme could be used under another theme, is it possible to locate the root project from multiple themes down? Can I also find the locations of all ancestor themes in between?

I suppose I could traverse up the directory structure myself but is there some other way? My idea for this theme is that it will search for component files in the parent themes with my own naming syntax, and I want to be able to control how the files override each other myself.

Edit:
I realize this is off-topic now so I've asked another question for this [here](https://github.com/gatsbyjs/gatsby/discussions/29504).
